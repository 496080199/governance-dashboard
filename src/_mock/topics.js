export default {
  mainnet: [
    {
      topic: 'Foundation Proposal',
      active: true,
      topic_blurb:
        'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
      date: '2018-03-12', //yyyy-mm-dd
      verified: true,
      submitted_by: {
        name: 'Dai Foundation',
        link: 'https://dai-dai-dai.now.sh/'
      },
      proposals: [
        {
          title:
            'Vote YES to the five core principles of the Maker Governance philosophy',
          proposal_blurb: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer vehicula sapien mi, vitae suscipit felis pellentesque a. Aenean condimentum vitae neque quis sagittis. Duis nec purus neque. Duis quis ultricies magna, in facilisis nisl. Suspendisse aliquet nisl et nisl mattis dictum. Donec varius sed lacus sit amet mollis. `,
          about: require('./demo.md'),
          source: '0x305505C8C9D51602f3Ebe0CE984aB2AEb0df3172',
          date: '2018-03-12', //yyyy-mm-dd
          verified: true,
          submitted_by: {
            name: 'Dai Foundation',
            link: 'https://dai-dai-dai.now.sh/'
          },
          documents: [
            {
              name: 'Risk...',
              link: 'https://...'
            }
          ]
        },
        {
          title:
            'Vote NO to the five core principles of the Maker Governance philosophy',
          proposal_blurb: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer vehicula sapien mi, vitae suscipit felis pellentesque a. Aenean condimentum vitae neque quis sagittis. Duis nec purus neque. Duis quis ultricies magna, in facilisis nisl. Suspendisse aliquet nisl et nisl mattis dictum. Donec varius sed lacus sit amet mollis. `,
          about: require('./demo.md'),
          source: '0xA9eB03F3a499E274170c378f958E4A0cBea2c7ef',
          date: '2018-03-12',
          verified: true,
          submitted_by: {
            name: 'Dai Foundation',
            link: 'https://dai-dai-dai.now.sh/'
          },
          documents: [
            {
              name: 'Risk...',
              link: 'https://...'
            }
          ]
        }
      ]
    },
    {
      topic: 'OSM, Debt Ceiling',
      active: false,
      topic_blurb:
        'With the introduction of the OSMs, we add an additional layer of security to our price feeds. ETH/USD and MKR/USD prices are now delayed by one hour, so we have time to react in the event the price feeds are compromised.',
      date: '2018-03-12', //yyyy-mm-dd
      verified: true,
      submitted_by: {
        name: 'Dai Foundation',
        link: 'https://dai-dai-dai.now.sh/'
      },
      proposals: [
        {
          title:
            'Raise the Debt Ceiling to $100M, add a 1 hour time delay to price feed changes',
          proposal_blurb: `Yesterday there was a sudden spike in CDP creation and we saw, in the span of just a few hours, the largest amount of Dai created in the shortest time in our history. A total of 7 Million Dai was drawn across a few CDPs representing a 16.27% increase in circulating Dai.`,
          about: require('./raise-roof.md'),
          source: '0x8ffb52208c08d4254e06737be0e1f0fe271b76b2',
          date: '2018-03-12',
          verified: true,
          submitted_by: {
            name: 'Dai Foundation',
            link: 'https://dai-dai-dai.now.sh/'
          },
          documents: [
            {
              name: 'Risk...',
              link: 'https://...'
            }
          ]
        }
      ]
    }
  ],
  kovan: [
    {
      topic: 'Foundation Proposal',
      active: true,
      topic_blurb:
        'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
      date: '2018-03-12', //yyyy-mm-dd
      verified: true,
      submitted_by: {
        name: 'Dai Foundation',
        link: 'https://dai-dai-dai.now.sh/'
      },
      proposals: [
        {
          title:
            'Vote YES to the five core principles of the Maker Governance philosophy',
          proposal_blurb: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer vehicula sapien mi, vitae suscipit felis pellentesque a. Aenean condimentum vitae neque quis sagittis. Duis nec purus neque. Duis quis ultricies magna, in facilisis nisl. Suspendisse aliquet nisl et nisl mattis dictum. Donec varius sed lacus sit amet mollis. `,
          about: require('./demo.md'),
          source: '0x0c0fC0952790A96D60CD82cA865C7bb1233477C3',
          date: '2018-03-12', //yyyy-mm-dd
          verified: true,
          submitted_by: {
            name: 'Dai Foundation',
            link: 'https://dai-dai-dai.now.sh/'
          },
          documents: [
            {
              name: 'Risk...',
              link: 'https://...'
            }
          ]
        },
        {
          title:
            'Vote NO to the five core principles of the Maker Governance philosophy',
          proposal_blurb: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer vehicula sapien mi, vitae suscipit felis pellentesque a. Aenean condimentum vitae neque quis sagittis. Duis nec purus neque. Duis quis ultricies magna, in facilisis nisl. Suspendisse aliquet nisl et nisl mattis dictum. Donec varius sed lacus sit amet mollis. `,
          about: require('./demo.md'),
          source: '0x66fF2801fDfEa24db2873abf977cA32226c175BD',
          date: '2018-03-12',
          verified: true,
          submitted_by: {
            name: 'Dai Foundation',
            link: 'https://dai-dai-dai.now.sh/'
          },
          documents: [
            {
              name: 'Risk...',
              link: 'https://...'
            }
          ]
        }
      ]
    }
  ]
};
