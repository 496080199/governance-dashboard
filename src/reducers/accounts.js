import uniqWith from 'ramda/src/uniqWith';
import concat from 'ramda/src/concat';
import pipe from 'ramda/src/pipe';
import differenceWith from 'ramda/src/differenceWith';
import pick from 'ramda/src/pick';

import { createReducer } from '../utils/redux';
import {
  getMkrBalance,
  getProxyStatus,
  getLinkedAddress,
  getVotingPower,
  getVotedSlate,
  getSlateAddresses,
  getNumDeposits
} from '../chain/read';
import { AccountTypes } from '../utils/constants';
import { add, eq, subtract, promisedProperties } from '../utils/misc';
import {
  SEND_MKR_TO_PROXY_SUCCESS,
  WITHDRAW_MKR_SUCCESS,
  INITIATE_LINK_REQUEST,
  APPROVE_LINK_SUCCESS
} from './proxy';
import { createSubProvider } from '../chain/hw-wallet';
import { netNameToId } from '../utils/ethereum';
import { toChecksum } from '../chain/web3';
import values from 'ramda/src/values';

// Constants ----------------------------------------------

const REMOVE_ACCOUNTS = 'accounts/REMOVE_ACCOUNTS';
const SET_ACTIVE_ACCOUNT = 'accounts/SET_ACTIVE_ACCOUNT';
const FETCHING_ACCOUNT_DATA = 'accounts/FETCHING_ACCOUNT_DATA';
export const UPDATE_ACCOUNT = 'accounts/UPDATE_ACCOUNT';
const ADD_ACCOUNT = 'accounts/ADD_ACCOUNT';
const SET_UNLOCKED_MKR = 'accounts/SET_UNLOCKED_MKR';
const FIND_HARDWARE_ACCOUNT = 'accounts/FIND_HARDWARE_ACCOUNT';
const FIND_HARDWARE_ACCOUNT_FAILURE = 'accounts/FIND_HARDWARE_ACCOUNT_FAILURE';
export const NO_METAMASK_ACCOUNTS = 'accounts/NO_METAMASK_ACCOUNTS';

// Selectors ----------------------------------------------

export function getAccount(state, address) {
  return state.accounts.allAccounts.find(a => eq(a.address, address));
}

export function getActiveAccount(state) {
  return getAccount(state, state.accounts.activeAccount);
}

export function getActiveVotingFor(state) {
  const activeAccount = getActiveAccount(state);
  return activeAccount ? activeAccount.votingFor : '';
}

export function activeCanVote(state) {
  const activeAccount = getActiveAccount(state);
  return (
    activeAccount &&
    activeAccount.hasProxy &&
    parseFloat(activeAccount.proxy.votingPower) > 0
  );
}

// Actions ------------------------------------------------

export const addAccounts = accounts => async dispatch => {
  dispatch({
    type: FETCHING_ACCOUNT_DATA,
    payload: true
  });
  for (let account of accounts) {
    const {
      hasProxy,
      type: proxyRole,
      address: proxyAddress
    } = await getProxyStatus(account.address);
    let currProposal = Promise.resolve('');
    if (hasProxy) {
      currProposal = currProposal.then(() =>
        Promise.all([
          getNumDeposits(proxyAddress),
          (async () => {
            const slate = await getVotedSlate(proxyAddress);
            const addresses = await getSlateAddresses(slate);
            return addresses[0] || '';
          })()
        ]).then(
          // NOTE for now we just take the first address in the slate since we're
          // assuming that they're only voting for one in the frontend. This
          // should be changed if that changes
          // also, if there's nothing locked in chief, we take this adddress as not voting for anything
          ([deposits, addresses]) => (Number(deposits) === 0 ? '' : addresses)
        )
      );
    }
    const _payload = {
      ...account,
      address: toChecksum(account.address),
      mkrBalance: getMkrBalance(account.address),
      hasProxy,
      proxyRole,
      votingFor: currProposal,
      proxy: (async () => ({
        address: hasProxy ? toChecksum(proxyAddress) : '',
        votingPower: hasProxy ? await getVotingPower(proxyAddress) : 0
      }))()
    };
    const fetchLinkedAccountData = async () => {
      if (hasProxy) {
        const otherRole = proxyRole === 'hot' ? 'cold' : 'hot';
        const linkedAddress = await getLinkedAddress(proxyAddress, otherRole);
        return {
          address: linkedAddress,
          proxyRole: otherRole,
          mkrBalance: await getMkrBalance(linkedAddress)
        };
      } else return {};
    };
    const [payload, linkedAccount] = await Promise.all([
      promisedProperties(_payload),
      fetchLinkedAccountData()
    ]);
    payload.proxy.linkedAccount = { ...linkedAccount };
    await dispatch({ type: ADD_ACCOUNT, payload });
  }
  return dispatch({ type: FETCHING_ACCOUNT_DATA, payload: false });
};

export const addAccount = account => async dispatch => {
  return dispatch(addAccounts([account]));
};

export const removeAccounts = accounts => ({
  type: REMOVE_ACCOUNTS,
  payload: accounts
});

export const updateAccount = account => ({
  type: UPDATE_ACCOUNT,
  payload: account
});

// After the initial load, this will generally be called when an account
// is selected in the account box dropdown
export const setActiveAccount = address => ({
  type: SET_ACTIVE_ACCOUNT,
  payload: address
});

export const getHardwareAccount = type => async (dispatch, getState) => {
  dispatch({ type: FIND_HARDWARE_ACCOUNT });
  const subprovider = createSubProvider(type, {
    networkId: netNameToId(getState().metamask.network),
    promisify: true
  });
  try {
    const addressesMap = await subprovider.getAccounts();
    const address = values(addressesMap)[0];
    dispatch(addAccount({ address, type, subprovider }));
  } catch (err) {
    console.error(err);
    dispatch({ type: FIND_HARDWARE_ACCOUNT_FAILURE, payload: err });
  }
};

// Reducer ------------------------------------------------

// Reducer helpers
const uniqByAddress = uniqWith((a, b) => a.address === b.address);
const uniqConcat = pipe(
  concat,
  uniqByAddress
);
const addressCmp = (x, y) => x.address === y.address;
const withUpdatedAccount = (accounts, updatedAccount) => {
  return accounts.map(
    account =>
      account.address === updatedAccount.address &&
      account.type === updatedAccount.type
        ? {
            ...account,
            ...updatedAccount
          }
        : account
  );
};

export const fakeAccount = {
  address: '0xbeefed1bedded2dabbed3defaced4decade5dead',
  type: 'fake',
  hasProxy: true,
  proxyRole: 'hot',
  votingFor: '0xbeefed1bedded2dabbed3defaced4decade5feed',
  mkrBalance: 333,
  proxy: {
    address: '0xproxyfake',
    votingPower: 111,
    linkedAccount: {
      address: '0xbeefed1bedded2dabbed3defaced4decade5feed',
      proxyRole: 'cold',
      mkrBalance: 222
    }
  }
};

const initialState = {
  activeAccount: '',
  fetching: true,
  // activeAccount: '0xbeefed1bedded2dabbed3defaced4decade5dead', // just for dev
  allAccounts: [
    // fakeAccount
  ]
};

const updateProxyBalance = adding => (state, { payload: amount }) => {
  let account = getActiveAccount({ accounts: state });
  if (!adding) {
    if (typeof amount === 'number') amount = -1 * amount;
    if (typeof amount === 'string') amount = '-' + amount;
  }

  account = {
    ...account,
    mkrBalance: subtract(account.mkrBalance, amount),
    proxy: {
      ...account.proxy,
      votingPower: add(account.proxy.votingPower, amount)
    }
  };

  let allAccounts = withUpdatedAccount(state.allAccounts, account);

  let linkedAccount = getAccount(
    { accounts: state },
    account.proxy.linkedAccount.address
  );
  if (linkedAccount) {
    linkedAccount = {
      ...linkedAccount,
      proxy: {
        ...linkedAccount.proxy,
        votingPower: add(linkedAccount.proxy.votingPower, amount)
      }
    };
    allAccounts = withUpdatedAccount(allAccounts, linkedAccount);
  }

  return { ...state, allAccounts };
};

const accounts = createReducer(initialState, {
  [REMOVE_ACCOUNTS]: (state, { payload: accounts }) => ({
    ...state,
    allAccounts: differenceWith(addressCmp, state.allAccounts, accounts)
  }),
  [UPDATE_ACCOUNT]: (state, { payload: updatedAccount }) => ({
    ...state,
    allAccounts: withUpdatedAccount(state.allAccounts, updatedAccount)
  }),
  [ADD_ACCOUNT]: (state, { payload: account }) => {
    if (!Object.keys(AccountTypes).includes(account.type)) {
      throw new Error(`Unrecognized account type: "${account.type}"`);
    }

    return {
      ...state,
      allAccounts: uniqConcat([account], state.allAccounts)
    };
  },
  [SET_ACTIVE_ACCOUNT]: (state, { payload: address }) => ({
    ...state,
    allAccounts: state.allAccounts,
    activeAccount: address
  }),
  [SET_UNLOCKED_MKR]: (state, { payload }) => ({
    ...state,
    activeAccountUnlockedMkr: payload.mkr
  }),
  [FETCHING_ACCOUNT_DATA]: (state, { payload }) => ({
    ...state,
    fetching: payload
  }),
  [NO_METAMASK_ACCOUNTS]: state => ({
    ...state,
    fetching: false
  }),
  [SEND_MKR_TO_PROXY_SUCCESS]: updateProxyBalance(true),
  [WITHDRAW_MKR_SUCCESS]: updateProxyBalance(false),
  [INITIATE_LINK_REQUEST]: (state, { payload }) => {
    const hotAccount = {
      ...getAccount({ accounts: state }, payload.hotAddress),
      proxyRole: 'hot'
    };

    const coldAccount = {
      ...getAccount({ accounts: state }, payload.coldAddress),
      proxyRole: 'cold'
    };

    return {
      ...state,
      allAccounts: withUpdatedAccount(
        withUpdatedAccount(state.allAccounts, hotAccount),
        coldAccount
      )
    };
  },

  [APPROVE_LINK_SUCCESS]: (state, { payload }) => {
    let hotAccount = getAccount({ accounts: state }, payload.hotAddress);
    let coldAccount = getAccount({ accounts: state }, payload.coldAddress);

    hotAccount = {
      ...hotAccount,
      proxy: {
        ...hotAccount.proxy,
        linkedAccount: pick(['address', 'mkrBalance', 'type'], coldAccount)
      }
    };

    coldAccount = {
      ...coldAccount,
      proxy: {
        ...coldAccount.proxy,
        linkedAccount: pick(['address', 'mkrBalance', 'type'], hotAccount)
      }
    };

    return {
      ...state,
      allAccounts: withUpdatedAccount(
        withUpdatedAccount(state.allAccounts, hotAccount),
        coldAccount
      )
    };
  }
});

export default accounts;
