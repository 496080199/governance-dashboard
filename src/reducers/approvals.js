import { createReducer } from '../utils/redux';
import { getApprovalCount } from '../chain/read';

// Constants ----------------------------------------------

const APPROVALS_REQUEST = 'voteTally/APPROVALS_REQUEST';
const APPROVALS_SUCCESS = 'voteTally/APPROVALS_SUCCESS';
const APPROVALS_FAILURE = 'voteTally/APPROVALS_FAILURE';

// Actions ------------------------------------------------

export const initApprovalsFetch = proposals => dispatch => {
  dispatch({ type: APPROVALS_REQUEST });
  Promise.all(
    proposals.map(({ source }) =>
      getApprovalCount(source).then(approvals => ({
        [source]: approvals
      }))
    )
  )
    .then(approvals => {
      const approvalsObj = {};
      for (let approval of approvals) {
        const [address, amt] = Object.entries(approval)[0];
        approvalsObj[address] = parseFloat(amt).toFixed(2);
      }
      dispatch({ type: APPROVALS_SUCCESS, payload: { approvalsObj } });
    })
    .catch(error => {
      // TODO: notify user or throw to a fallback component
      console.error(error);
      dispatch({ type: APPROVALS_FAILURE });
    });
};

// Reducer ------------------------------------------------

const initialState = {
  fetching: true,
  approvals: {}
};

const approvals = createReducer(initialState, {
  [APPROVALS_REQUEST]: state => ({
    ...state,
    fetching: true
  }),
  [APPROVALS_SUCCESS]: (_, { payload }) => ({
    fetching: false,
    approvals: payload.approvalsObj
  }),
  [APPROVALS_FAILURE]: state => ({
    ...state,
    fetching: false
  })
});

export default approvals;
