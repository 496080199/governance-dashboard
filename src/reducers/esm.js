import { createReducer } from '../utils/redux';

// Constants ----------------------------------------------

const ESM_STATE_REQUEST = 'esm/ESM_STATE_REQUEST';
const ESM_STATE_SUCCESS = 'esm/ESM_STATE_SUCCESS';
const ESM_STATE_FAILURE = 'esm/ESM_STATE_FAILURE';

// Actions ------------------------------------------------

export const esmInit = () => async dispatch => {
  dispatch({ type: ESM_STATE_REQUEST });
  try {
    const esmService = await window.maker.service('esm');
    const totalStaked = await esmService.getTotalStaked();
    const canFire = await esmService.canFire();
    const thresholdAmount = await esmService.thresholdAmount();
    const fired = await esmService.fired();
    const cageTime = await window.maker
      .service('smartContract')
      .getContract('END')
      .when();
    dispatch({
      type: ESM_STATE_SUCCESS,
      payload: { totalStaked, canFire, thresholdAmount, fired, cageTime }
    });
  } catch (err) {
    dispatch({ type: ESM_STATE_FAILURE });
  }
};

// Reducer ------------------------------------------------

const initialState = {
  fetching: true,
  canFire: false,
  thresholdAmount: 0,
  totalStaked: 0,
  fired: false,
  cageTime: 0
};

const esm = createReducer(initialState, {
  [ESM_STATE_REQUEST]: () => ({
    ...initialState,
    fetching: true
  }),
  [ESM_STATE_SUCCESS]: (_, { payload }) => ({
    ...payload,
    fetching: false
  }),
  [ESM_STATE_FAILURE]: state => ({
    ...state,
    fetching: false
  })
});

export default esm;
