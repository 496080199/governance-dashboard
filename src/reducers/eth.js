import { createReducer } from '../utils/redux';
import maker from '../chain/maker';

// Constants ----------------------------------------------

const ETH_INFO_REQUEST = 'eth/ETH_INFO_REQUEST';
const ETH_INFO_SUCCESS = 'eth/ETH_INFO_SUCCESS';
const ETH_INFO_FAILURE = 'eth/ETH_INFO_FAILURE';

// Actions ------------------------------------------------

export const ethInit = () => async dispatch => {
  try {
    dispatch({ type: ETH_INFO_REQUEST });
    const ethPrice = await maker.service('price').getEthPrice();
    maker.service('web3')._web3.eth.getGasPrice((error, gasPrice) => {
      if (error) throw new Error('unable to fetch current gas price');
      dispatch({
        type: ETH_INFO_SUCCESS,
        payload: {
          price: ethPrice.toNumber(),
          gasCost: gasPrice.shift(-18).toFixed()
        }
      });
    });
  } catch (err) {
    dispatch({ type: ETH_INFO_FAILURE });
  }
};

// Reducer ------------------------------------------------

const initialState = {
  fetching: false,
  gasCost: '',
  price: ''
};

const eth = createReducer(initialState, {
  [ETH_INFO_REQUEST]: state => ({
    ...state,
    fetching: true
  }),
  [ETH_INFO_SUCCESS]: (state, { payload }) => ({
    ...state,
    gasCost: payload.gasCost,
    price: payload.price
  }),
  [ETH_INFO_FAILURE]: () => ({
    ...initialState
  })
});

export default eth;
