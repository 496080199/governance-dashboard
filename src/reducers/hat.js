import { createReducer } from '../utils/redux';
import { getHat } from '../chain/read';

// Constants ----------------------------------------------

const HAT_REQUEST = 'hat/HAT_REQUEST';
const HAT_SUCCESS = 'hat/HAT_SUCCESS';
const HAT_FAILURE = 'hat/HAT_FAILURE';

// Actions ------------------------------------------------

export const hatInit = () => dispatch => {
  dispatch({ type: HAT_REQUEST });
  getHat()
    .then(address => {
      dispatch({ type: HAT_SUCCESS, payload: { address } });
    })
    .catch(() => {
      // notify user of error or throw to fallback. Network disconnect?
      dispatch({ type: HAT_FAILURE });
    });
};

// Reducer ------------------------------------------------

const initialState = {
  fetching: false,
  hatAddress: ''
};

const hat = createReducer(initialState, {
  [HAT_REQUEST]: state => ({
    ...state,
    fetching: true
  }),
  [HAT_SUCCESS]: (_, { payload }) => ({
    hatAddress: payload.address,
    fetching: false
  }),
  [HAT_FAILURE]: () => ({
    addresses: '',
    fetching: false
  })
});

export default hat;
