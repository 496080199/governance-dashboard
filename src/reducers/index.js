import { combineReducers } from "redux";
import metamask from "./metamask";
import modal from "./modal";
import topics from "./topics";
import tally from "./tally";
import approvals from "./approvals";
import proxy from "./proxy";
import user from "./user";
import accounts from "./accounts";

const rootReducer = combineReducers({
  metamask,
  modal,
  topics,
  tally,
  approvals,
  proxy,
  user,
  accounts
});

export default rootReducer;
