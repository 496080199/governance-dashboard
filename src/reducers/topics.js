/////////////////////////////////////////////////
//// Mocked backend w/ the current proposals ////
/////////////////////////////////////////////////

import { createReducer } from '../utils/redux';
import { initApprovalsFetch } from './approvals';
import mocked from '../_mock/topics';
import { getEtchedSlates } from '../chain/read';

// Constants ----------------------------------------------

const TOPICS_SUCCESS = 'toics/TOPICS_SUCCESS';

// Actions ------------------------------------------------

export const topicsInit = network => async dispatch => {
  if (network === 'ganache') {
    // look up all slates
    const slates = await getEtchedSlates();
    const topics = [
      {
        topic: 'Test topic',
        proposals: slates.map(source => ({
          title: 'Test proposal',
          source
        }))
      }
    ];
    dispatch({ type: TOPICS_SUCCESS, payload: topics });
  } else {
    dispatch({ type: TOPICS_SUCCESS, payload: mockedBackend[network] });
  }
  dispatch(initApprovalsFetch());
};

// Reducer ------------------------------------------------

const mockedBackend = mocked;

const topics = createReducer([], {
  [TOPICS_SUCCESS]: (_, { payload }) => payload
});

export default topics;
