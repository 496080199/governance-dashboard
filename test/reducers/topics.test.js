import { topicsInit } from '../../src/reducers/topics';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

test('topicsInit returns a TOPICS_SUCCESS action with topics as a payload', async () => {
  const store = mockStore({});
  store.dispatch(topicsInit('mainnet'));
  expect(store.getActions()).toMatchSnapshot();
});
