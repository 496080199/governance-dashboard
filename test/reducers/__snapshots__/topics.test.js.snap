// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`topicsInit returns a TOPICS_SUCCESS action with topics as a payload 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "active": true,
        "date": "2018-03-12",
        "end_timestamp": 1537027200000,
        "govVote": true,
        "id": 1,
        "proposals": Array [
          Object {
            "about": "test-file-stub",
            "date": "2018-03-12",
            "documents": Array [
              Object {
                "link": "https://...",
                "name": "Risk...",
              },
            ],
            "end_timestamp": 1537027200000,
            "proposal_blurb": "Voting for this proposal means voting YES to the five core principles of the Maker Governance philosophy as <a target=\\"_blank\\" href=\\"https://medium.com/makerdao/foundation-proposal-v2-f10d8ee5fe8c\\"> updated on Medium</a> on August 21, 2018",
            "source": "0x305505C8C9D51602f3Ebe0CE984aB2AEb0df3172",
            "submitted_by": Object {
              "link": "https://dai-dai-dai.now.sh/",
              "name": "Dai Foundation",
            },
            "title": "Vote YES to the five core principles of the Maker Governance philosophy",
            "verified": true,
          },
          Object {
            "about": "test-file-stub",
            "date": "2018-03-12",
            "documents": Array [
              Object {
                "link": "https://...",
                "name": "Risk...",
              },
            ],
            "end_timestamp": 1537027200000,
            "proposal_blurb": "Voting for this proposal means voting NO to the five core principles of the Maker Governance philosophy as <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://medium.com/makerdao/foundation-proposal-v2-f10d8ee5fe8c\\"> updated on Medium</a> on August 21, 2018",
            "source": "0xA9eB03F3a499E274170c378f958E4A0cBea2c7ef",
            "submitted_by": Object {
              "link": "https://dai-dai-dai.now.sh/",
              "name": "Dai Foundation",
            },
            "title": "Vote NO to the five core principles of the Maker Governance philosophy",
            "verified": true,
          },
        ],
        "submitted_by": Object {
          "link": "https://dai-dai-dai.now.sh/",
          "name": "Dai Foundation",
        },
        "topic": "Foundation Proposal Governance Vote",
        "topic_blurb": "Here you can vote YES or NO to the proposed five core principles of the Maker Governance philosophy as <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://medium.com/makerdao/foundation-proposal-v2-f10d8ee5fe8c\\"> updated on Medium</a> on August 21, 2018.
        This type of vote is a <strong style=\\"font-weight:bold;font-style:oblique;\\">Governance Vote</strong>, and its objective is to represent resolution on a matter or collection of matters. This vote will not result in changes to the Dai Credit System (to change the state of the system, for example to change the debt ceiling, a second type of vote called an <strong style=\\"font-weight:bold;font-style:oblique;\\">Executive Vote</strong> is used).",
        "verified": true,
      },
    ],
    "type": "toics/TOPICS_SUCCESS",
  },
]
`;
